language/diem-vm/parallel-executor/src/outcome_array.rs

```diff
@@ -21,54 +21,29 @@ impl<T: Default, E: Send> OutcomeArray<T, E> {
+        // a block here. And each txn id will be scheduled to execute exactly once.
 
         let entry = &self.results[idx];
         assert!(entry.set(res).is_ok());
     }
 
     pub fn get_all_results(self, valid_length: usize) -> Result<Vec<T>, E> {
         let len = self.results.len();
-        let mut results = self
-            .results
-            .into_iter()
-            .take(valid_length)
-            .map(|cell| cell.into_inner())
-            .collect::<Option<Vec<_>>>()
-            .ok_or(Error::InvariantViolation)?;
-
-        match results.pop() {
-            Some(ExecutionStatus::Success(output)) => {
-                // Last transaction is executed successfully, meaning that all transactions outputs will be returned.
-                assert_eq!(len, valid_length);
-                let mut other_outputs = results
-                    .into_iter()
-                    .map(|status| match status {
-                        ExecutionStatus::Success(output) => Ok(output),
-                        _ => Err(Error::InvariantViolation),
-                    })
-                    .collect::<Result<Vec<_>, _>>()?;
-                other_outputs.push(output);
-                Ok(other_outputs)
-            }
-            Some(ExecutionStatus::SkipRest(output)) => {
-                // Last transaction calls to skip the trailing transactions, thus we will fill the
-                // rest of the transaction outputs with default values
-                let mut other_outputs = results
-                    .into_iter()
-                    .map(|status| match status {
-                        ExecutionStatus::Success(output) => Ok(output),
-                        _ => Err(Error::InvariantViolation),
-                    })
-                    .collect::<Result<Vec<_>, _>>()?;
-                other_outputs.push(output);
-                other_outputs.resize_with(len, Default::default);
-                Ok(other_outputs)
-            }
-            // Last transaction triggered an error. Drop the entire execution and return the error.
-            Some(ExecutionStatus::Abort(err)) => Err(err),
-            Some(ExecutionStatus::Retry(_)) => Err(Error::InvariantViolation),
-            None => Ok(vec![]),
+        let mut final_results = Vec::with_capacity(valid_length);
+        for (idx, status) in self.results.into_iter().take(valid_length).enumerate() {
+            let t = match status.into_inner() {
+                Some(ExecutionStatus::Success(t)) => t,
+                Some(ExecutionStatus::SkipRest(t)) if idx == valid_length - 1 => t,
+                Some(ExecutionStatus::SkipRest(_)) => return Err(Error::InvariantViolation),
+                Some(ExecutionStatus::Abort(err)) => return Err(err),
+                Some(ExecutionStatus::Retry(_)) => return Err(Error::InvariantViolation),
+                _ => return Err(Error::InvariantViolation),
```

Comment:
isn't this just None? Might want to write it out to avoid pain if you ever refactor the enum